MBS-12552 (II): Refactor CoreEntity in data layer

yvanzo opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Server.
    We appreciate your time and interest in helping our project!

    Please use this template to help us review your change.

    Depending on your change, some sections may be unneeded, just remove these.
    For example, small pull requests usually don’t need the section “Action”.

    Remember that the more helpful info your pull request includes,
    the easier it is for us to understand and review your changes.

    Ensure that you’ve read through and followed the Contributing Guidelines, at
    https://github.com/metabrainz/musicbrainz-server/blob/master/CONTRIBUTING.md
-->

# Problem
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

As part of MBS-12552, it appears that the current class `CoreEntity` in data layer isn’t about relationships, it is rather mostly about GID/MBID.

# Solution
<!--
    Talk about technical details, considerations, or other interesting points.
-->

* Split `CoreEntitty` class into roles, `GIDRedirect`, `MainTable`, `Name`, `Relatable` in the data layer
* Rename `CoreEntittyCache` role to `GIDEntityCache` in the data layer

# Notes

* Having moved from a class to a role makes mandatory to add explicit `requires`.

# Open questions

* Should the `CoreEntity` entity class be renamed as `PermanentEntity` too? It involves undoing some changes from <https://github.com/metabrainz/musicbrainz-server/pull/2681>.
  - Pros:
    * Restore a close mapping between Entity classes and Data classes.
    * Free the term “core” to keep its use for public domain data as in [MusicBrainz Database](https://musicbrainz.org/doc/MusicBrainz_Database) / [Download](https://musicbrainz.org/doc/MusicBrainz_Database/Download) and for table list in [`MusicBrainz::Server::Constants`](https://github.com/metabrainz/musicbrainz-server/blob/v-2022-09-06/lib/MusicBrainz/Server/Constants.pm#L509-L732).
  - Cons:
    * More refactoring (but there will be more anyway).
    * Break with NGS usage in code (but not in public documentation).

reosarevok commented:
I think doing the same in `Entity::` makes sense - whatever we decide to keep we should keep with the same meaning in both layers IMO.

What do you mean with this?

> Break with NGS usage in code (but not in public documentation).

You mean that we have stuff like `CoreEntityT` in JS? We should probably be refactoring those too anyway, whether now or a bit later.

reosarevok commented on t/lib/t/MusicBrainz/Server/Data/PermanentEntityCache.pm:
```diff
@@ -31,14 +31,14 @@ with 't::Context' => { -excludes => '_build_context' };
-    extends 't::CoreEntityCache::MyEntityData';
-    with 'MusicBrainz::Server::Data::Role::CoreEntityCache';
+    extends 't::PermanentEntityCache::MyEntityData';
+    with 'MusicBrainz::Server::Data:Role::PermanentEntityCache';
```
```suggestion
    with 'MusicBrainz::Server::Data::Role::PermanentEntityCache';
```

yvanzo commented:
> I think doing the same in `Entity::` makes sense

I agree but mainly asking to @mwiencek because it somehow gets back on [his comment to the ticket](https://tickets.metabrainz.org/browse/MBS-12552?focusedCommentId=64394&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-64394) I tried to follow so far (with [my answer](https://tickets.metabrainz.org/browse/MBS-12552?focusedCommentId=64402&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-64402) and my previous pull request).

> What do you mean with this?
> 
> > Break with NGS usage in code (but not in public documentation).

It mostly refers to the conversation we had with @mwiencek on `#metabrainz` yesterday, which stated that the term “core” has been (partly) defined when moving to NGS.

> stuff like `CoreEntityT` in JS? We should probably be refactoring those too anyway, whether now or a bit later.

I’m starting with the oldest piece of code (assuming it is the most reliable) and incrementally consolidate terms and definitions.
I just updated MBS-12552 with “Permanent” (entity) suggestion.

reosarevok commented:
This should ideally at least also rename the `Data::CoreEntity` and `Data::CoreEntityCache` test files/packages :)

mwiencek commented:
Regarding the name, "permanent" makes me think of something that can't be removed.  `GidEntity` (globally-identifiable entity) or `RedirectableGidEntity` might be more to-the-point?

Of course, there are some entities that use gids but don't support redirects, which is why I think separate Gid/GidRedirect roles might be nice.

> It uses a lot of stuff from Entity so moving it to separated roles would need a lot of explicit requires.

I think 3-5 requires per file isn't bad and might actually make the code clearer.

> Should the CoreEntity entity class be renamed as PermanentEntity too?

Well as you know, I still like "core entity" for referring to the relatable entities. :)  In fact I would like keeping `Data::CoreEntity` and using it to combine the gid, gid redirect, name, and id roles, and then merging linkable into it, since linkable is not used anywhere outside of the core entity classes.  (I consider it to be a bug or implementation quirk that Collection inherits from CoreEntity, and I would change that.)  This class could also be named `Data::RelatableEntity`, which I wouldn't object to.  I just prefer keeping existing terms where possible.

yvanzo commented:
> This should ideally at least also rename the `Data::CoreEntity` and `Data::CoreEntityCache` test files/packages :)

It’s already the case.

yvanzo commented:
> Regarding the name, "permanent" makes me think of something that can't be removed.

Which is mostly correct as we prefer merging rather than removing these entities when possible.

> `GidEntity` (globally-identifiable entity) or `RedirectableGidEntity` might be more to-the-point?

Ok, I can try that. I personally find those hideous but at least it helps with moving on. :)

> Of course, there are some entities that use gids but don't support redirects, which is why I think separate Gid/GidRedirect roles might be nice.

It makes sense.
 
> > It uses a lot of stuff from Entity so moving it to separated roles would need a lot of explicit requires.
> 
> I think 3-5 requires per file isn't bad and might actually make the code clearer.

Ok, I can do that too, or at least will open a separate draft pull request if it isn’t strictly needed.
To be honest, I also tried to minimize the needed refactoring as well, but I’m not opposed to a deeper refactoring.

> > Should the CoreEntity entity class be renamed as PermanentEntity too?
> 
> Well as you know, I still like "core entity" for referring to the relatable entities. :)

It seems more to be about nostalgia than rationale though. :) Unfortunately it would unnecessarily waste a term that could be usefully replaced by “relatable” which is more explanatory, leaving the term available for another of its already current usages (which are numerous). See also the 2nd point of above Pros for Permanent (which are Cons for Core here).

> In fact I would like keeping `Data::CoreEntity` and using it to combine the gid, gid redirect, name, and id roles, and then merging linkable into it, since linkable is not used anywhere outside of the core entity classes.

Ok, I will keep this name for now, even though I’m pretty sure that we will have to reconsider this decision later on.
I will focus on splitting into roles for now which is definitively a move forward.

> (I consider it to be a bug or implementation quirk that Collection inherits from CoreEntity, and I would change that.)

I agree, even though addressing implementation quirks is not really the main goal here.

> This class could also be named `Data::RelatableEntity`, which I wouldn't object to. I just prefer keeping existing terms where possible.

Ok, I will keep it as an option. Time for coding.

yvanzo commented:
@mwiencek: Please let me know if you agree with the changes so far before I go ahead with further changes? Some progress has been made; The code works but some tests have not been updated yet. I completely moved the code of `Data::CoreEntity` to separate roles then dropped it entirely as it was empty. It allows to make explicitly require `_hash_to_row` and `_type` for (old and new) roles.

mwiencek reviewed (commented):
Overall the changes look good, though for "WIP: Drop Data::CoreEntity empty package" I'd probably still keep some type of class (or even a role) that composes the shared roles used by relatable entity types.

reosarevok commented on lib/MusicBrainz/Server/Data/Role/EntityModelClass.pm:
```diff
@@ -0,0 +1,42 @@
+package MusicBrainz::Server::Data::Role::EntityModelClass;
```
Requiring every entity to be in entities.json would seem good to me. I recently fixed a warning because of track there missing mbid.relatable entirely, so I'd actually even prefer us to have a fairly strict model of how entities need to look like and what they should have (set to false if they don't actually have it). But making sure they're all there would be a good start.

yvanzo replied:
@mwiencek: I don’t see the added value of requiring this definition to be in JSON if it isn’t consumed outside of `MusicBrainz::Server::Data`.

@reosarevok: It seems absurd to do that: There are 302 entity types in this JSON file and 32 different possible subkeys; that would make 9664 keys, paradoxically for no value either.

Having a documentation for this JSON file would be more helpful.

reosarevok commented:
This might already be in the plans, but just in case: `t::MusicBrainz::Server::Data::CoreEntity` should be renamed, and `t::GIDEntityCache::MyEntityData` still does `extends 'MusicBrainz::Server::Data::CoreEntity'` which unsurprisingly fails the test - it's probably extend `Data::Entity` with `Data::Role::GetByGID` now? Unless we keep a class/role for relatable types, but tbh even if we do, this should probably work with the basic non-relatable GID entities too anyway and should use the basic roles.

yvanzo commented on lib/MusicBrainz/Server/Data/Role/GID.pm:
```diff
@@ -0,0 +1,58 @@
+package MusicBrainz::Server::Data::Role::GID;
```
No, it is used independently by `LinkAttributeType` and `LinkType` classes.

mwiencek replied:
I think that ideally anything with a GID should be able to use `Data::Role::GID`, but we could look into that later.

yvanzo commented on lib/MusicBrainz/Server/Data/Role/GIDRedirect.pm:
```diff
@@ -1,24 +1,35 @@
-extends 'MusicBrainz::Server::Data::Entity';
-with 'MusicBrainz::Server::Data::Role::GetByGID';
+requires '_main_table', '_type';
+requires 'c', 'get_by_gid', 'get_by_gids', 'get_by_id', 'get_by_ids', 'sql';
```
No, following the comment above.

yvanzo commented:
@reosarevok wrote: _This might already be in the plans_

It was in the [above comment](https://github.com/metabrainz/musicbrainz-server/pull/2706#issuecomment-1285891710): _some tests have not been updated yet_.
Now updated.

yvanzo commented:
@mwiencek wrote: _keep some type of class (or even a role) that composes the shared roles used by relatable entity types_

Done with a new role `Relatable` in commit <https://github.com/metabrainz/musicbrainz-server/pull/2706/commits/6647f3a85eb86bfdbcec41722c2ce8f521239c93>.

mwiencek reviewed (approved):
This feels like an improvement to me.  Thanks a lot for working on it!

yvanzo commented:
Updated the PR description and the ticket (by removing “Permanent”), added commit https://github.com/metabrainz/musicbrainz-server/pull/2706/commits/8f7ed607f108ffdba2f243e6c9cc3207ed090bbc for an oversight.

Thanks for your support, will keep it up.
